### PR TITLE
@vcr decorator for unit/doctests - recording/playback of network traffic via python sockets

### PR DESCRIPTION
**Pull request by barsch**
### What does this PR do?

Any Python socket communication in unittests (decorated with the ```@vcr``` function) and/or doctests (containing a ```# doctest: +VCR```) will be recorded on the first run and saved into a special 'vcrtapes' directory as single pickled file for each test case. Future test runs will reuse those recorded network session allowing for faster tests without any network connection. In order to create a new recording one just needs to remove/rename the pickled session file(s).

So pretty similar to [VCRPy](https://github.com/kevin1024/vcrpy) but on socket level instead of HTTP (request, urllib3, etc.) - so therefore it should be more powerful ...

### Why was it initiated?  Any relevant issues?

Network tests tend to fail sporadically, need usually a very long time (compared to other tests) and (surprise!)  require a network connection (if not mocked).

This PR tackles all three issues mentioned above. Here some performance tests (network connection had been disabled during tests of the vcr branch):

```
(Python36) d:\Workspace\obspy>git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.

(Python36) d:\Workspace\obspy>obspy-runtests -d obspy.clients.syngine.tests.test_client
Running d:\workspace\obspy\obspy\scripts\runtests.py, ObsPy version '1.0.2.post0+902.gbfe6b3c90a.obspy.master'
..............
----------------------------------------------------------------------
Ran 14 tests in 9.743s
OK

(Python36) d:\Workspace\obspy>git checkout vcr
Switched to branch 'vcr'
Your branch is up-to-date with 'origin/vcr'.

(Python36) d:\Workspace\obspy>obspy-runtests -d obspy.clients.syngine.tests.test_client
Running d:\workspace\obspy\obspy\scripts\runtests.py, ObsPy version '1.0.2.post0+903.g151e12679f.obspy.vcr'
..............
----------------------------------------------------------------------
Ran 14 tests in 0.432s
OK
```

### ToDo
- [X] ensure platform compatibility
  - [X] Win (https://ci.appveyor.com/project/obspy/vcr/build/1.0.31)
  - [X] Linux (https://travis-ci.org/obspy/vcr/builds/210161304)
  - [X] Mac (https://travis-ci.org/obspy/vcr/builds/210161304)
- [x] clients.arclink: issues on Linux with current telnet implementation - I may have to rewrite that :/
- [x] clients.seishub: needs vcrtapes for client doctests (@megies)
- [x] clients.seedlink: needs reworking - current tests don't work - can't add @vcr therefore ...
- [x] clients.fdsn: could cover more tests - currently the service discovery thread (which is using sockets too) is intercepting occasionally the test case itself (@krischer)
- [x] ensure the whole test suite works without network connection
- [x] remove (non-)network tests options from obspy-runtests - all tests should be run as default setting
- [X] SSLSocket support (HTTPS)
- [X] no arclink specific hacks anymore -> more general option for recv_timeout and recv_endmarker
- [X] needs documentation
- [X] needs `CHANGELOG.txt` entry
- [x] move it into an extra Python package as its highly reusable - also to increase test coverage by using all appveyor env instead only selected ones due to huge number of tests we have in obspy -> see https://github.com/obspy/vcr/ - you need to ```pip install -U vcr``` on your local env
- [ ] include vcr as a git submodule in core/util/testing ?
- [ ] squash old vcr tapes in the git history

### CODE REVIEW

**Comment by megies** [comment]
+TESTS:ALL

**Comment by megies** [comment]
I just wanted to provide vcrtapes for seishub, but it seems seishub client was broken by #1325..

**Comment by megies** [comment]
The networking stuff of `Stream` test case `test_read` should be refactored out into a separate test `test_read_with_url_via_network`. I would even do it in maintenance branch, but I don't want to get you into problems with rebasing..

**Comment by megies** [comment]
@barsch, I pushed a commit with vcrtapes for seishub, no idea why the doctest vcrtapes weren't recorded..

Those vcrtapes were recorded with this commit added on top of this branch: https://github.com/obspy/obspy/pull/1664/commits/dd8238cfdb680057f19b2a07ee5098e32605e187

**Comment by barsch** [comment]
@megies - doctests are currently monkey patched only via ```obspy-runtests``` script and not at module level - so run ```obspy-runtests -d clients.seishub``` and it should work (using ```#doctests: +VCR``` comment once somewhere within a single doctest)

**Comment by megies** [comment]
Hmm.. somehow `seishub/__init__.py` doctests are not run on master or vcr branch. They are run on maintenance_1.0.x though...

**Comment by barsch** [comment]
so its already an issue on master ...

**Comment by barsch** [comment]
clients.seishub works for me on windows except for test_localcache as its not been decorated - any reason for that?

**Comment by megies** [comment]
> clients.seishub works for me on windows except for test_localcache as its not been decorated - any reason for that?

Ah, my bad, will add it soonish..

**Comment by megies** [comment]
@barsch added missing unittest vcrtape for seishub. but can't add doctest vcrtape because that test is failing, see #1664.

**Comment by megies** [comment]
@barsch.. there's some issues with vcr.py highlighted by our flake8 check, see CircleCI fail log:
 
```python
  280 class VCR(object):
  281     """
  282     """
  283     @classmethod
  284     def reset(cls, debug=False):
  285         global vcr_playlist, vcr_status, vcr_debug
  286         vcr_playlist = []
  287         vcr_status = VCR_RECORD
  288         vcr_debug = debug
>>289         vcr_recv_timeout = 2
>>290         vcr_recv_endmarker = None
>>291         vcr_recv_size = None
  292 
```

The last three variables are not declared global so these statements are without effect..

```python
  364             # check for arclink module
  365             if 'arclink' in source_filename:
  366                 vcr_recv_endmarker = 'END\r\n'
  367             else:
>>368                 vcr_recv_endmarker = ''
```

This variable is not used inside this scope and also not declared global, so this statement seems to be without effect.

**Comment by barsch** [comment]
yeah thanks for poiting out - its pretty much WIP and I needed to sleep - so I accidently pushed it isntead of local commit only ...

however I'm not a fan of globals at all - any better idea for a more elegant solution - having anything within the decorator as before works without globals but is pretty much untestable ...

**Comment by barsch** [comment]
ok this became now a new module see https://github.com/obspy/vcr - primary because its much faster testable than via an obspy branch and so it can be reused in other projects  - download it via ```pip install -U vcr``` if you want to use it! Please use this repository also for non-obspy specific feedback on vcr.

This PR still valid as it integrates vcr into obspy - its so far looking pretty good ...

**Comment by megies** [comment]
Did all seishub vcr tapes from scratch.

This branch will need some cleanup before merging, once it's ready. Let me know when everything is done, I can rebase and squash then.

**Comment by megies** [comment]
Seishub is fully vcr-ified now, verified on Py27 and Py36, tests run in under one second with network unplugged.

**Comment by megies** [comment]
If I remove the threading during service discovery and do a plain sequential download of wadls, then FDSN can be fully vcr-ified and tests run without network access in under 3 seconds.

I know that the threaded wadl download is a slick feat, but maybe it's worth to lose it to ease up testing? I tried and connecting to IRIS with plain sequential download takes me 5.3 seconds (averaged across 10 connections), compared to ~8 seconds with the current threaded wadl download.. and connecting should be a one-off operation in most practical use cases?

**Comment by barsch** [comment]
at least an option to temporary disable it would be cool

**Comment by megies** [comment]
Hmm.. but if we do this and use it in tests than the threaded download would be untested code.. :-/

**Comment by krischer** [comment]
We could add a slight delay for each thread so that the download start order would be deterministic. It would not really impact total runtime but it should be enough for the vcr decorator to do its thing. Would that work?

**Comment by krischer** [comment]
The threading is really useful for very slow datacenters were it improves things a lot. Sometimes one or more of the fdsn services also time-out while one works. In these cases the threading is a lot fsater.

**Comment by barsch** [comment]
I wouldn't want to remove it - just an option to disable/preload without threads - some code which I can put into TestCase.setUp to run service discovery only once at my desire and disable it for the actually test cases

**Comment by krischer** [comment]
That would be easy enough - the results are already cached - just fill it before you run the tests: https://github.com/obspy/obspy/blob/master/obspy/clients/fdsn/client.py#L108

**Comment by megies** [comment]
> We could add a slight delay for each thread so that the download start order would be deterministic

So, essentially a short `time.sleep()` after `thread.start()`? Tried it but didn't work with vcr. Patching `self.services` is not a real solution since it can't cover our doctests.. One option would be an `Client.__init__(..., threaded_wadl_downloads=True)` kwarg, but then again, if we switch it to False in our tests that means that our default client behavior is untested..

**Comment by megies** [comment]
@barsch do you have an overview of the current status on this one? Can we get it ready by Sunday? Any help needed?

**Comment by barsch** [comment]
- current test cases in io.seedlink not working - therefore I can't vcr them
- not sure about io.fdsn - don't know where I can savely disable service discovery and where not - probably need to split up into two test files - one which can be used with vcr and one which can be optional disabled for environment without network connection

everything should be covered with current release of vcr (0.0.5) - found some non-trivial issue with requests package on Linux resulting into a failing vcr master - but that shouldn't effect the current test cases in ObsPy

**Comment by megies** [comment]
Rebased on current master and force-pushed.

Also always switched on ALL modules for runtests on Appveyor+Travis via the ymls.

**Comment by megies** [comment]
Appveyor runs ALL tests in 5mins, only broken seedlink tests failing: http://tests.obspy.org/75341/
:-)

**Comment by megies** [comment]
We should add another option to `obspy-runtests` to skip using vcr tapes, something like `obspy-runtests --no-vcr`.

Also, we should maybe deprecate `--all` and always run all test suites.. ?

**Comment by megies** [comment]
OK, so there's basically two things to resolve and then VCR should be ready..
 - #1733 
 - threaded fdsn service discovery not vcr compatible

CC @krischer @barsch

**Comment by megies** [comment]
For me FDSN tests are running now with only one exception: `test_redirection_auth` because it explicitly wipes the service discovery cache internally and thus forces fresh service discovery.

Also, since with VCR we are now fully circumventing service discovery, we need to test it explicitlely without VCR (as the current threaded approach can not be used with VCR). CC @krischer

**Comment by megies** [comment]
I've updated our vcr test setup, we can now use two separate decorators (provided in `obspy.core.util.decorator`) in our unittests:
 - `@network_test`: to mark test cases that always will be executed via live internet (if available, otherwise skipped)
 - `@vcr`: to mark test cases that use pre-recorded network interaction and with option `--no-vcr` get executed over live internet connection (or are skipped if no internet connection is available)

See https://github.com/obspy/obspy/pull/1663/commits/670fc1506ddea5e3b64a9844018307c4020b3e32:

This means that:
 * with default settings and internet available:
     - tests marked with `@vcr` will use pre-recorded vcrtapes
     - tests marked with `@network_test` will be executed with live
       internet
 * with default settings and *no* internet available:
     - tests marked with `@vcr` will use pre-recorded vcrtapes
     - tests marked with `@network_test` will be skipped with reason 'no
       internet connection (...)'
 * with option `--no-vcr` and internet available:
     - tests marked with `@vcr` will be skipped
     - tests marked with `@network_test` will be executed with live
       internet
 * with option `--no-vcr` and *no* internet available:
     - tests marked with `@vcr` will be skipped
     - tests marked with `@network_test` will be skipped with reason 'no
       internet connection (...)'

## Example:

Note that fdsn test suite below has 2 tests marked as `@network_test` and
several others marked as `@vcr`.

### Internet connected:

this runs 2 networked tests and many tests via vcr tapes
```
$ obspy-runtests clients.fdsn
...................................................................................................
----------------------------------------------------------------------
Ran 99 tests in 15.773s
OK
```

this runs all tests via live internet
(therefore slowest but tests against real/live production servers)
```
$ obspy-runtests clients.fdsn --no-vcr
...................................................................................................
----------------------------------------------------------------------
Ran 99 tests in 74.078s
OK
```

### Internet disconnected:


this skips 2 networked tests and runs other tests via vcr tapes
```
$ obspy-runtests clients.fdsn
................................ss.................................................................
----------------------------------------------------------------------
Ran 99 tests in 2.040s
OK
```

this skips 2 networked tests and also skips vcr tests
(since --no-vcr was specified but no internet is available)
```
$ obspy-runtests clients.fdsn --no-vcr
.ssssss..s.....s.s..s.s.ssss.sssss....ss...........................................................
----------------------------------------------------------------------
Ran 99 tests in 1.565s
OK
```

**Comment by megies** [comment]
Rebasing on current master fixing some conflicts and force-pushing. Previous head was https://github.com/obspy/obspy/pull/1663/commits/c4594a49fad7d3880ab2565b0905ab8c9702be95.

**Comment by megies** [comment]
Rebased and force-pushed again

**Comment by megies** [comment]
Introduced an option `--offline` to force-skip any tests that would use the internet (at least if they're marked as `+VCR`/`+NETWORK_TEST` for doctests or `@vcr`/`@network_test` for unittests)

**Comment by megies** [comment]
So #1733 is the only thing that's holding up this PR now. :-)

**Comment by megies** [comment]
screenshot from 2017-03-30 16 29 13

---

screenshot from 2017-03-30 16 31 22

**Comment by barsch** [comment]
now we only need a global option to overwrite all vcr tests ... :)

**Comment by megies** [comment]
> now we only need a global option to overwrite all vcr tests ... :)

you mean force recording of new vcr tapes?

**Comment by barsch** [comment]
> you mean force recording of new vcr tapes?

essentially by setting ```from vcr import VCRSystem; VCRSystem.overwrite = True```, btw. instead of having your own vcr decorator ```from vcr import VCRSystem; VCRSystem.disabled= True``` would have worked too

**Comment by megies** [comment]
> btw. instead of having your own vcr decorator from vcr import VCRSystem; VCRSystem.disabled= True would have worked too for disabling vcr mechanics completely

Nah.. still need a wrapper around it because a VCR test will fall back to a network test if VCR is disabled, and those will be skipped without network. Just disabling VCR for them would execute them using internet, no matter if it's available or not or whether internet is force-disabled.

**Comment by megies** [comment]
Rebased + force-pushed

**Comment by megies** [comment]
Here's the first fully vcr-ified test run with LAN cable unplugged: http://tests.obspy.org/76315/

**Comment by megies** [comment]
We were discussing, probably including obspy/vcr as a git submodule in core/util/testing/ is an idea to avoid the additional external dependency for testing..

Also we need to cleanup history a bit (at least squash old vcr tapes out of the PR), will od it tomorrow or so..

**Comment by megies** [comment]
This might be a Python 3->2 / future artifact? https://ci.appveyor.com/project/obspy/obspy/build/1.0.4581-master/job/8fobkif99dxlt3mr#L911

**Comment by krischer** [comment]
> We were discussing, probably including obspy/vcr as a git submodule in core/util/testing/ is an idea to avoid the additional external dependency for testing..

Let's do a subtree - way easier to clone.

**Comment by krischer** [comment]
> This might be a Python 3->2 / future artifact? https://ci.appveyor.com/project/obspy/obspy/build/1.0.4581-master/job/8fobkif99dxlt3mr#L911

Does it also happen with vcr off?

**Comment by megies** [comment]
> Let's do a subtree - way easier to clone.

Didn't know that one.. sounds good from a first glimpse over the documentation! :-)

**Comment by megies** [comment]
> Does it also happen with vcr off?

Huh.. dunno decided this is not worth much thought and just replaced these parameter types with "normal"/intuitive names  (unicode -> str , long -> int)

**Comment by megies** [comment]
Hmm.. strange, this one if failing withing vcr decorator.. https://ci.appveyor.com/project/obspy/obspy/build/1.0.4582-master/job/ap3ef8qh2m3qjsxg#L741

**Comment by megies** [comment]
Python 3.3 seems to not work: https://travis-ci.org/obspy/obspy/jobs/217222457#L710
Wild guess is that `future` might make problems for double-underscore namespace-mangling in Py 3.3???

**Comment by megies** [comment]
> Let's do a subtree - way easier to clone.

Hmm.. even with subtree we'd need some manual tinkering (like [here](https://jrsmith3.github.io/merging-a-subdirectory-from-another-repo-via-git-subtree.html)) to get rid of the top level directory and then some more to add `future` imports to `__init__.py` in vcr.

Dunno, maybe just copy vcr files manually and commit them..?

**Comment by krischer** [comment]
> Dunno, maybe just copy vcr files manually and commit them..?

Works for me. I guess it will also not be updated that often.

**Comment by krischer** [comment]
Too lazy to read all the commits - do we need to squash this to get rid of eventually committed multiple tapes?

**Comment by megies** [comment]
Yes, we need to get rid of old tapes, I will do the clean up soonish.

**Comment by megies** [comment]
Rebased on current master, gonna start cleaning up history now. Current head is ce24a6fb2c36702b0c73443bc67ed71e769ec23a just in case I mess up.. ;-)

**Comment by barsch** [comment]
> > Dunno, maybe just copy vcr files manually and commit them..?
> Works for me. I guess it will also not be updated that often.

Please don't - either use subtree/submodule or leave it as it is as cheap (python-only, small size) dependency - having the source in multiple locations is not good practice

we have a lot dependencies which are way bigger and less used ...

**Comment by megies** [comment]
> we have a lot dependencies which are way bigger and less used ...

sure.. but it might be hard to control that it gets installed (e.g. in user installations) so that running tests is possible.. dunno

**Comment by megies** [comment]
All vcrtapes together are 18MB.. but I guess that's the price to pay..

**Comment by krischer** [comment]
> All vcrtapes together are 18MB.. but I guess that's the price to pay..

Puh...that is really a tough pill to swallow. I think we can reduce the load from the worst offenders - there should be no need for them to download so much stuff.

```
-- The five worst offenders make up almos 13 MB.
3.0M clients/seishub/tests/vcrtapes/test_client.test_get_preview.vcr
2.6M clients/earthworm/tests/vcrtapes/__init__.__init__.vcr
2.6M clients/earthworm/tests/vcrtapes/test_client.test_availability.vcr
2.6M clients/earthworm/tests/vcrtapes/client.get_availability.vcr
1.9M clients/seishub/tests/vcrtapes/test_client.test_get_preview_by_ids.vcr

-- 3 MB more if all files larger then 100K are taken into account.
546K clients/seishub/tests/vcrtapes/test_client.test_get_waveform_with_metadata.vcr
452K clients/syngine/tests/vcrtapes/client.get_waveforms_bulk.vcr
430K clients/iris/tests/vcrtapes/test_client.test_resp.vcr
386K clients/neic/tests/vcrtapes/test_client.test_get_waveform_nscl.vcr
312K clients/syngine/tests/vcrtapes/__init__.__init__.vcr
270K clients/syngine/tests/vcrtapes/client.get_waveforms.vcr
264K clients/iris/tests/vcrtapes/test_client.test_evalresp.vcr
181K clients/fdsn/tests/vcrtapes/test_client.test_iris_example_queries_dataselect.vcr
171K clients/arclink/tests/vcrtapes/test_client.test_get_inventory_twice.vcr
121K clients/seishub/tests/vcrtapes/__init__.__init__.vcr
```

**Comment by barsch** [comment]
I'll add gzip compression to vcr - will reduce filesize by ca 50% - however rest must be done in tests

**Comment by megies** [comment]
Size reduction by compression is much more than 50% it seems to me. 👍 

Will work on those oversized tests as well..

**Comment by megies** [comment]
```
-- The five worst offenders make up almos 13 MB.

3.0M clients/seishub/tests/vcrtapes/test_client.test_get_preview.vcr
 -> shrinked

2.6M clients/earthworm/tests/vcrtapes/__init__.__init__.vcr
 -> can be marked +NETWORK_TEST instead, functionality is tested by test_client.test_availability

2.6M clients/earthworm/tests/vcrtapes/test_client.test_availability.vcr
 -> nothing we can do right now.. we would need to make MENUSCNL command working

2.6M clients/earthworm/tests/vcrtapes/client.get_availability.vcr
 -> can be marked +NETWORK_TEST instead, functionality is tested by  test_client.test_availability

1.9M clients/seishub/tests/vcrtapes/test_client.test_get_preview_by_ids.vcr
 -> shrinked


-- 3 MB more if all files larger then 100K are taken into account.
546K clients/seishub/tests/vcrtapes/test_client.test_get_waveform_with_metadata.vcr
 -> not much that can be done, this internally does quite a few small requests to seishub, multiple metadata requests
452K clients/syngine/tests/vcrtapes/client.get_waveforms_bulk.vcr
430K clients/iris/tests/vcrtapes/test_client.test_resp.vcr
386K clients/neic/tests/vcrtapes/test_client.test_get_waveform_nscl.vcr
312K clients/syngine/tests/vcrtapes/__init__.__init__.vcr
270K clients/syngine/tests/vcrtapes/client.get_waveforms.vcr
264K clients/iris/tests/vcrtapes/test_client.test_evalresp.vcr
181K clients/fdsn/tests/vcrtapes/test_client.test_iris_example_queries_dataselect.vcr
171K clients/arclink/tests/vcrtapes/test_client.test_get_inventory_twice.vcr
121K clients/seishub/tests/vcrtapes/__init__.__init__.vcr
```

**Comment by megies** [comment]
Hmmm.. just noticed.. one thing that is sorely missing in VCR yet is checking what we send to the server..  Currently the code of the test routine can be changed and the test is still passing because the VCR tape does not assert the outgoing traffic is what it's supposed to be.. :-/

**Comment by barsch** [comment]
as soon you change something either in the test itself or in the underlying client source code you have to recreate the vcr file - it really only records the incoming traffic

**Comment by megies** [comment]
I don't think there'a way around checking/asserting outgoing traffic as well. Right now our code base could be broken while vcr tests still pass (e.g. one of our clients producing a malformed request -- vcr would still playback the assumed response from the server leading to a passing test).

And even if we check outgoing traffic, I think we need to assert that the test code hasn't changed as well, otherwise we could still produce the same outgoing traffic while calling completely different code (although I agree that such a breach is far less likely to happen).

bottom line: IMHO vcr has to check outgoing traffic otherwise we can't rely on it for our testing.

**Comment by megies** [comment]
I'll push a proposal for outgoing traffic assertions in vcr in a short while..

**Comment by megies** [comment]
Only a few more glitches left to tackle.. -_-
 - Win Py2: https://ci.appveyor.com/project/obspy/obspy/build/1.0.4647-master/job/rm61xwnyui1g1sck#L719
 - Py 3.3: https://travis-ci.org/obspy/obspy/jobs/218574897
 - and one SSL error for the `read()` with our https://examples.obspy.org, seems to affect newer Debuntus but only 64bit: http://tests.obspy.org/?version=1.0.3.post0%2B1053.ge01d364ec6.obspy.vcr&node=docker-, e.g. http://tests.obspy.org/77260/#1

note to self.. one more http header to quench.. `Content-Type: application/x-www-form-urlencoded`  e.g. http://tests.obspy.org/77270/#2

**Comment by megies** [comment]
@barsch can you reproduce and try to debug [this one](https://ci.appveyor.com/project/obspy/obspy/build/1.0.4647-master/job/rm61xwnyui1g1sck#L719) locally on Win Py 2.7 (installing vcr from that open PR branch)?

**Comment by barsch** [comment]
```@vcr(debug=True)``` fails with
```
Traceback (most recent call last):
  File "D:\Workspace\obspy\obspy\clients\seedlink\tests\test_basic_client.py", line 26, in <module>
    class ClientTestCase(unittest.TestCase):
  File "D:\Workspace\obspy\obspy\clients\seedlink\tests\test_basic_client.py", line 30, in ClientTestCase
    @vcr(debug=True)
TypeError: vcr() got an unexpected keyword argument 'debug'
```

so I had to hack the decorator in and got:

```
VCR PLAYBACK (test_get_waveform) ...
Loaded playlist:
SHA256: acae40974d86a3edcf479302f88ed05896a371c5cefc7b75df932b12faefd28b
  0: connect ((u'rtserver.ipgp.fr', 18000),) {}
  1: send ('HELLO\r',) {}
  2: recv (1024,) {}
  3: send ('STATION  FDF G\r',) {}
  4: recv (1024,) {}
  5: send ('SELECT 00LHN\r',) {}
  6: recv (1024,) {}
  7: send ('TIME 2017,4,3,12,29,37 2017,4,3,12,29,57\r',) {}
  8: recv (1024,) {}
  9: send ('END\r',) {}
 10: recv (8192,) {}
 11: recv (8184,) {}
 12: connect ((u'rtserver.ipgp.fr', 18000),) {}
 13: send ('HELLO\r',) {}
 14: recv (1024,) {}
 15: send ('STATION  CLF G\r',) {}
 16: recv (1024,) {}
 17: send ('SELECT 00BHZ\r',) {}
 18: recv (1024,) {}
 19: send ('TIME 2017,4,3,12,29,37 2017,4,3,12,29,47\r',) {}
 20: recv (1024,) {}
 21: send ('END\r',) {}
 22: recv (8192,) {}
 23: recv (8184,) {}
SHA256 sum of source code matches playlist
   __init__ 2 1 0 None
   setsockopt (65535, 4098, 65536) {}
   setsockopt (65535, 4, 1) {}
   setsockopt (65535, 8, 1) {}
   connect ((u'rtserver.ipgp.fr', 18000),) {}  |  connect ((u'rtserver.ipgp.fr', 18000),) {} -> None
  checking:  connect ((u'rtserver.ipgp.fr', 18000),) {}  |  connect ((u'rtserver.ipgp.fr', 18000),) {}
  checking, after normalization:  connect ((u'rtserver.ipgp.fr', 18000),) {}  |  connect ((u'rtserver.ipgp.fr', 18000),) {}
   close
E
======================================================================
ERROR: test_get_waveform (__main__.ClientTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "<decorator-gen-54>", line 2, in test_get_waveform
  File "D:\Workspace\obspy\obspy\core\util\decorator.py", line 86, in vcr
    vcr_module.vcr(func, debug=True)(*args, **kwargs)
  File "d:\workspace\vcr\vcr\core.py", line 637, in _vcr_inner
    value = func(*args, **kwargs)
  File "D:\Workspace\obspy\obspy\clients\seedlink\tests\test_basic_client.py", line 37, in test_get_waveform
    st = self.client.get_waveforms(*request)
  File "D:\Workspace\obspy\obspy\clients\seedlink\basic_client.py", line 125, in get_waveforms
    self._slclient.run(packet_handler=self._packet_handler)
  File "D:\Workspace\obspy\obspy\clients\seedlink\slclient.py", line 244, in run
    slpack = self.slconn.collect()
  File "D:\Workspace\obspy\obspy\clients\seedlink\client\seedlinkconnection.py", line 822, in collect
    self.connect()
  File "D:\Workspace\obspy\obspy\clients\seedlink\client\seedlinkconnection.py", line 1074, in connect
    raise SeedLinkException(msg % (e))
SeedLinkException: u"cannot connect to SeedLink server: u'socket connect time-out 4.0s'"

----------------------------------------------------------------------
Ran 1 test in 4.010s

FAILED (errors=1)
```

however the current debug output has changed - and I have no clue how to read it yet

**Comment by megies** [comment]
Rebased and force-pushed so that it's mergeable into master and appveyor can run..

**Comment by megies** [comment]
> however the current debug output has changed - and I have no clue how to read it yet

so, it only executes the first command (the connect) end then exits somehow..

**Comment by megies** [comment]
Hmm.. lots of fdsn fails now, with the fdsn service discovery cache pickled on Python 3.3 (trying to get rid of problems on Python 3.3, when pickling on Python 3.6).

So, I'll revert and go back to pickling on Python 3.6..

**Comment by megies** [comment]
@barsch, regarding that failing seedlink test on Win Py2.7.. if you record a vcr tape on Win Py3 and playback on Win Py2.7.. does that work?

**Comment by barsch** [comment]
@megies: same result

how do you recreate vcr files effectively: currently ```@vcr(overwrite=True)``` raises an error and deleting the vcr file complains that there is no tape file - I had to hack again to get it to work ...

**Comment by megies** [comment]
> how do you recreate vcr files effectively: currently @vcr(overwrite=True) raises an error and deleting the vcr file complains that there is no tape file - I had to hack again to get it to work ...

```
$ obspy-runtests -h
[...]
Networking Options:
  --no-vcr              do not use pre-recorded vcr tapes with network
                        interactions for network tests. using this option,
                        tests that use vcr tapes will use internet connection
                        if possible or they will be skipped if no internet
                        connection is available.
  --vcr-record          record new vcr tapes. by default this will not
                        overwrite existing tapes but instead use existing
                        tapes for playback. use option "--vcr-overwrite" to
                        overwrite existing tapes.
  --vcr-overwrite       overwrite existing vcr tapes (has no effect without "
                        --vcr-record").
  --offline             do not use internet connection for network tests even
                        if available, i.e. force-skip tests that would use an
                        internet connection.
[...]
```

So.. either delete individual tape files and `obspy-runtests [module] --vcr-record` or overwrite all tape files with `obspy-runtests [module] --vcr-record --vcr-overwrite`.

**Comment by megies** [comment]
Locally run docker tests are here: http://tests.obspy.org/?git=336e9e4c62&node=docker-

**Comment by megies** [comment]
@barsch regarding the SSL problem (in obspy/core/tests/vcrtapes/test_stream.test_read_url_via_network.vcr)

On Python 2 traffic looks like: (it initializes a `Socket` object, and not a `SSLSocket` like on Python 3!)
```
   vcr_getaddrinfo ('examples.obspy.org', 443, 0, 1) {} [(10, 1, 6, '', ('2a01:238:426b:7300:f053:c7ab:4e68:7416', 443, 0, 0)), (2, 1, 6, '', ('85.214.244.127', 443))]
   __init__ 10 1 6 None
   setsockopt (6, 1, 1) {}
   settimeout (None,) {}
   connect (('2a01:238:426b:7300:f053:c7ab:4e68:7416', 443, 0, 0),) {} None
   fileno () {} -> 12
   settimeout (None,) {}
   close
```
On Python 3 traffic looks like:
```
   vcr_getaddrinfo ('examples.obspy.org', 443, <AddressFamily.AF_UNSPEC: 0>, <SocketKind.SOCK_STREAM: 1>) {} [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('2a01:238:426b:7300:f053:c7ab:4e68:7416', 443, 0, 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('85.214.244.127', 443))]
   __init__ AddressFamily.AF_INET6 SocketKind.SOCK_STREAM 6 None
   setsockopt (6, 1, 1) {}
   settimeout (None,) {}
   connect (('2a01:238:426b:7300:f053:c7ab:4e68:7416', 443, 0, 0),) {} None
   __init__ () {'server_side': False, 'do_handshake_on_connect': True, 'suppress_ragged_eofs': True, 'server_hostname': 'examples.obspy.org', '_context': <ssl.SSLContext object at 0x7fe4a89a0438>, '_session': None}
   getpeercert () {} {'subject': ((('commonName', 'ccoverage.obspy.org'),),), 'issuer': ((('countryName', 'US'),), (('organizationName', "Let's Encrypt"),), (('commonName', "Let's Encrypt Authority X3"),)), 'version': 3, 'serialNumber': '03A77965B30A172B0C7D6709406CDB0F73D6', 'notBefore': 'Jan 16 13:01:00 2017 GMT', 'notAfter': 'Apr 16 13:01:00 2017 GMT', 'subjectAltName': (('DNS', 'ccoverage.obspy.org'), ('DNS', 'deb-unstable.obspy.org'), ('DNS', 'deb.obspy.org'), ('DNS', 'docs.obspy.org'), ('DNS', 'docsets.obspy.org'), ('DNS', 'examples.obspy.org'), ('DNS', 'gallery.obspy.org'), ('DNS', 'obspy.org'), ('DNS', 'tests.obspy.org'), ('DNS', 'tutorial.obspy.org'), ('DNS', 'www.obspy.org')), 'OCSP': ('http://ocsp.int-x3.letsencrypt.org/',), 'caIssuers': ('http://cert.int-x3.letsencrypt.org/',)}
   sendall (b'GET /test.sac HTTP/1.1\r\nHost: examples.obspy.org\r\nUser-Agent: ObsPy/1.0.3.post0+1090.g943b85004a.dirty.obspy.vcr (Linux-4.9.0-0.bpo.2-amd64-x86_64-with-debian-8.7, Python 3.6.0)\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n',) {} None
   settimeout (None,) {}
   makefile ('rb',) {} <_io.BufferedReader name=7>
   close
```

**Comment by megies** [comment]
Remaining problems:
 - [ ] SSL problems on some 64bit Python 2.7 (Linux only?) systems: e.g. [here](http://tests.obspy.org/77743/)
 - [ ] [seedlink connection timeout in appveyor](https://ci.appveyor.com/project/obspy/obspy/build/1.0.4654-master/job/xp55tv1k79q11yxv#L735)
 - [ ] [Python 3.3 problems](https://travis-ci.org/obspy/obspy/jobs/218962139#L698) (seems to be not really vcr related but some pickle issues..)

I'll switch on debug output in CI for the first two temporarily..

**Comment by megies** [comment]
Docker test results: http://tests.obspy.org/?git=9bf769ccde&node=docker-

**Comment by megies** [comment]
Rebased and force-pushed for fresh CI results

**Comment by megies** [comment]
Rebased and force-pushed.

Deactivated all live network use in CI during tests now, so those tests that are not run through VCR but marked `@network_test`/`+NETWORK_TEST` are skipped in CI now.

We should add a switch in CI scripts to turn on live network tests for release candidate version numbers maybe..

**Comment by megies** [comment]
Rebased and force-pushed for fresh CI

**Comment by megies** [comment]
no idea why appveyor is hickupping with io.arclink??

https://ci.appveyor.com/project/obspy/obspy/build/1.0.4673-master/job/n873g5k5xtyb62hk#L509

all other tests pass on appveyor..

**Comment by megies** [comment]
Punting to 1.2.0.. otherwise we'll never even get to 1.1.0

Everybody feel free to tackle it of course..
